File: backend/app/schemas/vlm.py
"""
VLM 엔진 전용 요청·응답 스키마.

- VlmOcrResponse: 전체 텍스트 추출 (기존 OCR 호환 + confidence/bbox 추가)
- SchemaExtractResponse: Schema 기반 Key-Value 구조화 추출
- QaResponse: 문서 Q&A
"""
from __future__ import annotations

from pydantic import BaseModel, Field


# ── 공통 ──────────────────────────────────────────────

class BoundingBox(BaseModel):
    x: float
    y: float
    width: float
    height: float


# ── OCR ───────────────────────────────────────────────

class VlmOcrItem(BaseModel):
    text: str
    confidence: float | None = None
    bbox: BoundingBox | None = None


class VlmOcrResponse(BaseModel):
    success: bool = True
    model_id: str
    elapsed_ms: int = 0
    items: list[VlmOcrItem] = Field(default_factory=list)
    full_text: str = ""
    error: str | None = None
    prompt_mode: str | None = None  # auto | bbox | custom
    prompt_label: str | None = None  # 실제 사용 단계 (spotting, custom 등)
    raw_response_preview: str | None = None  # 모델 원본 응답 앞부분 (디버그)


# ── Schema 추출 ───────────────────────────────────────

class SchemaField(BaseModel):
    key: str
    description: str = ""
    type: str = "text"


class SchemaExtractItem(BaseModel):
    key: str
    value: str
    confidence: float | None = None
    bbox: BoundingBox | None = None


class SchemaExtractRequest(BaseModel):
    schema_fields: list[SchemaField]


class SchemaExtractResponse(BaseModel):
    success: bool = True
    model_id: str
    elapsed_ms: int = 0
    items: list[SchemaExtractItem] = Field(default_factory=list)
    error: str | None = None


# ── Q&A ───────────────────────────────────────────────

class QaResponse(BaseModel):
    success: bool = True
    model_id: str
    elapsed_ms: int = 0
    answer: str = ""
    confidence: float | None = None
    error: str | None = None


# ── 모델 목록 ─────────────────────────────────────────

class VlmModelInfo(BaseModel):
    model_id: str
    name: str
    description: str
    vram_gb: float
    loaded: bool = False


class VlmModelsResponse(BaseModel):
    models: list[VlmModelInfo] = Field(default_factory=list)
    current_model: str | None = None

File: frontend/src/types/vlm.ts
/* VLM 전용 타입 정의 */

export interface BoundingBox {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface VlmOcrItem {
  text: string;
  confidence?: number | null;
  bbox?: BoundingBox | null;
}

export type VlmOutputFormat = "bbox" | "text_only";

export interface VlmOcrOptions {
  outputFormat?: VlmOutputFormat;
}

export interface VlmRequestOptions {
  outputFormat?: VlmOutputFormat;
}

export interface VlmOcrResponse {
  success: boolean;
  model_id: string;
  elapsed_ms: number;
  items: VlmOcrItem[];
  full_text: string;
  error?: string | null;
  prompt_mode?: string | null;
  prompt_label?: string | null;
  output_format?: string | null;
  raw_response_preview?: string | null;
  model_raw_preview?: string | null;
}

/** exact_text: key = 이미지에 있는 문자열 그대로 | 그 외: key = 결과 필드 ID */
export type SchemaMatchMode = "exact_text" | "script_filter" | "semantic_field";

export type SchemaScriptFilter =
  | "han"
  | "hangul"
  | "latin"
  | "english"
  | "digit";

export interface SchemaField {
  key: string;
  /** location_hint (exact/script) 또는 의미 설명 (semantic) */
  description: string;
  type: string;
  match_mode?: SchemaMatchMode;
  script?: SchemaScriptFilter;
}

export interface SchemaExtractItem {
  key: string;
  value: string;
  confidence?: number | null;
  bbox?: BoundingBox | null;
}

export interface SchemaExtractResponse {
  success: boolean;
  model_id: string;
  elapsed_ms: number;
  items: SchemaExtractItem[];
  output_format?: string | null;
  error?: string | null;
}

export interface VlmModelInfo {
  model_id: string;
  name: string;
  description: string;
  vram_gb: number;
  loaded: boolean;
}

export interface VlmModelsResponse {
  models: VlmModelInfo[];
  current_model: string | null;
}

export type VlmMode = "ocr" | "schema" | "qa";

export interface QaResponse {
  success: boolean;
  model_id: string;
  elapsed_ms: number;
  answer: string;
  confidence?: number | null;
  output_format?: string | null;
  error?: string | null;
}
